File: app/src/main/java/com/bopr/android/smailer/ui/BatteryOptimizationHelper.java
package com.bopr.android.smailer.ui;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.PowerManager;

import androidx.appcompat.app.AlertDialog;

import com.bopr.android.smailer.R;

import static android.content.Context.POWER_SERVICE;
import static android.provider.Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS;

class BatteryOptimizationHelper {

    static boolean isIgnoreBatteryOptimizationRequired(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            String packageName = context.getApplicationContext().getPackageName();
            PowerManager pm = (PowerManager) context.getSystemService(POWER_SERVICE);
            return pm != null && !pm.isIgnoringBatteryOptimizations(packageName);
        }
        return false;
    }

    static void requireIgnoreBatteryOptimization(final Context context) {
        if (isIgnoreBatteryOptimizationRequired(context)) {
            new AlertDialog.Builder(context)
                    .setTitle("Battery optimization")
                    .setMessage(R.string.battery_optimization_reason)
                    .setNegativeButton(android.R.string.cancel, null)
                    .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                        Intent intent = new Intent();
                        intent.setAction(ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS);
                        context.startActivity(intent);
                    })
                    .show();
        }
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/ui/BatteryOptimizationHelper.kt
package com.bopr.android.smailer.ui

import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.PowerManager
import android.provider.Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS
import androidx.fragment.app.FragmentActivity
import com.bopr.android.smailer.R

/**
 * Helper to check battery optimization state and to request to ignore it.
 */
internal object BatteryOptimizationHelper {

    const val BATTERY_OPTIMIZATION_DIALOG_TAG = "battery-optimization-do-not-ask-again"

    fun isIgnoreBatteryOptimizationRequired(context: Context): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val packageName = context.applicationContext.packageName
            val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
            return !pm.isIgnoringBatteryOptimizations(packageName)
        }
        return false
    }

    fun requireIgnoreBatteryOptimization(activity: FragmentActivity) {
        if (isIgnoreBatteryOptimizationRequired(activity)) {
            showDialog(activity)
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private fun showDialog(activity: FragmentActivity) {
        activity.run {
            ConfirmCheckDialog(
                    title = getString(R.string.battery_optimization),
                    message = getString(R.string.battery_optimization_reason),
                    positiveButtonText = getString(R.string.proceed),
                    dialogTag = BATTERY_OPTIMIZATION_DIALOG_TAG) {
                startActivity(Intent(ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS))
            }.show(this)
        }
    }
}